File: Packages/NoteroMidiPlugin/Runtime/Native/midiDeviceAdapter/AdapterRegister.kt
package co.notero.midiplugin.midiDeviceAdapter

import android.media.midi.MidiDeviceInfo
import co.notero.midiplugin.NoteroMidiDevice
import co.notero.midiplugin.NoteroMidiInputPort

typealias DeviceAssertion = (
    midiDeviceInfo: MidiDeviceInfo,
    inputPort: NoteroMidiInputPort,
) -> NoteroMidiDevice?

class AdapterRegister {
    companion object {
        private val assertors = arrayListOf<DeviceAssertion>(
            GZUTPianoMidiDeviceAdapter.deviceAssertion,
            PopPianoMidiDeviceAdapter.deviceAssertion,
            DreamSASMidiDeviceAdapter.deviceAssertion,
            GenericPianoMidiDeviceAdapter.deviceAssertion,
        )

        fun assertDevice(
            midiDeviceInfo: MidiDeviceInfo,
            inputPort: NoteroMidiInputPort,
        ): NoteroMidiDevice? {
            for (assertor in assertors) {
                val device = assertor(midiDeviceInfo, inputPort)
                if (device != null) {
                    return device;
                }
            }
            return null
        }
    }
}

File: Packages/NoteroMidiPlugin/Runtime/Windows/MidiDeviceAdapter/AdapterRegister.cs
using System.Collections.Generic;
using System.Linq;
#if !UNITY_ANDROID || UNITY_EDITOR
using NAudio.Midi;
#endif

#if !UNITY_ANDROID || UNITY_EDITOR
namespace Notero.MidiPlugin.Windows.MidiDeviceAdapter
{
    public delegate MidiDevice DeviceAssertion(MidiOutCapabilities midiDeviceInfo, MidiOut midiOut);

    public static class AdapterRegister
    {
        private static List<DeviceAssertion> m_Assertors = new List<DeviceAssertion>()
        {
            GZUTPianoMidiDeviceAdapter.DeviceAssertion,
            PopPianoMidiDeviceAdapter.DeviceAssertion,
            DreamSASMidiDeviceAdapter.DeviceAssertion,
            GenericPianoMidiDeviceAdapter.DeviceAssertion
        };

        public static MidiDevice AssertDevice(MidiOutCapabilities midiDeviceInfo, MidiOut midiOut)
        {
            return m_Assertors.Select(assertor => assertor(midiDeviceInfo, midiOut)).FirstOrDefault(device => device != null);
        }
    }
}
#endif
